src/python/pants/engine/README.md

```diff
@@ -135,6 +167,18 @@ The initial Nodes are [launched by the engine](https://github.com/pantsbuild/pan
 but the rest of execution is driven by Nodes recursively calling `Context.get` to request their
 dependencies.
 
+### Registering Rules
+
+Currently, it is only possible to load rules into the pants scheduler in two ways: by importing and using them in `src/python/pants/bin/engine_initializer.py`, or by adding them to the list returned by a `rules()` method defined in `src/python/backend/<backend_name>/register.py`. Plugins cannot add new rules yet. Unit tests, however, can mix in `SchedulerTestBase` from `tests/python/pants_test/engine/scheduler_test_base.py` to generate and execute a scheduler from a given set of rules.
+
+In general, there are three types of rules you can define:
+
+1. an `@rule`, which has a single product type and selects its inputs as described above.
+2. a `SingletonRule`, which matches a product type with a value so the type can then be `Select`ed in an `@rule`.
+3. a `RootRule`, which declares a type that can be used as a *subject*, which means it can be provided as an input to a `product_request()` or a `Get` statement.
```

Comment:
So, the reason `RootRule` exists is only for the former case: `Scheduler.product_request`, et al. Those represent the root of a request. In the absence of `RootRules`, any subject type that was _only_ involved in a request "at runtime" (ie via `product_request`), would show up as an an unused or impossible path in the rulegraph, because no `@rule` could provide that type. `RootRule` allows for plugging that hole in rulegraph verification by saying: "yea, no other `@rule` provides this type, but trust me: someone might request one at runtime".

Another potential name for `RootRule` might be "ParamRule", or something similar... because it's essentially saying that the type represents a sort of "public API entrypoint" via a `product_request`. I'm fairly confident that `RootRule` is a necessary API, but the name is not at all final... Suggestions welcome!

----

You do not need to use `RootRule` if any `Get` requests exist in the graph, because the `Get` requests are statically verified in the graph. We know before runtime that they might be requested.

Reply:
I pilfered most of that comment into a few sentences in 5032fe9.

Reply:
`ParamRule` would be a great name, and thinking it over, the only problem is that `RootRule` is also a good name, for different reasons.